packages/playground/data-liberation/src/import/WP_Import_Session.php

```diff
@@ -0,0 +1,406 @@
+<?php
+
+/**
+ * Manages import session data in the WordPress database.
+ *
+ * Each import session is stored as a post of type 'import_session'.
+ * Progress, stage, and other metadata are stored as post meta.
+ */
+class WP_Import_Session {
+	const POST_TYPE = 'import_session';
+	/**
+	 * @TODO: Make it extendable
+	 * @TODO: Reuse the same entities list as WP_Stream_Importer
+	 */
+	const PROGRESS_ENTITIES = array(
+		'site_option',
+		'user',
+		'category',
+		'tag',
+		'term',
+		'post',
+		'post_meta',
+		'comment',
+		'comment_meta',
+		'file',
+	);
+	private $post_id;
+	private $cached_stage;
+
+	/**
+	 * Creates a new import session.
+	 *
+	 * @param array $args {
+	 *     @type string $data_source     The data source (e.g. 'wxr_file', 'wxr_url', 'markdown_zip')
+	 *     @type string $source_url      Optional. URL of the source file for remote imports
+	 *     @type int    $attachment_id   Optional. ID of the uploaded file attachment
+	 *     @type string $file_name       Optional. Original name of the uploaded file
+	 * }
+	 * @return WP_Import_Model|WP_Error The import model instance or error if creation failed
+	 */
+	public static function create( $args ) {
+		// Validate the required arguments for each data source.
+		// @TODO: Leave it up to filters to make it extendable.
+		switch ( $args['data_source'] ) {
+			case 'wxr_file':
+				if ( empty( $args['file_name'] ) ) {
+					_doing_it_wrong(
+						__METHOD__,
+						'File name is required for WXR file imports',
+						'1.0.0'
+					);
+					return false;
+				}
+				break;
+			case 'wxr_url':
+				if ( empty( $args['source_url'] ) ) {
+					_doing_it_wrong(
+						__METHOD__,
+						'Source URL is required for remote imports',
+						'1.0.0'
+					);
+					return false;
+				}
+				break;
+			case 'markdown_zip':
+				if ( empty( $args['file_name'] ) ) {
+					_doing_it_wrong(
+						__METHOD__,
+						'File name is required for Markdown ZIP imports',
+						'1.0.0'
+					);
+					return false;
+				}
+				break;
+		}
+
+		$post_id = wp_insert_post(
```

Comment:
Using custom type posts is a great idea. 👍

Reply:
Thank you! I was looking for a way to reuse as much of what we already have as possible. A custom table crossed my mind, and we still might need one [for the vector clock](https://core.trac.wordpress.org/ticket/60375#comment:24), but for managing metadata post types and meta seem perfect.

Reply:
Yes. In the near future, we will probably need a place to save binary data and similar data. But now, using custom types posts for this is perfectly fine.